### PR TITLE
fix(types): accept `$nested.syntax$` in WhereAttributeHash

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [n/a] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Tiny change that has become possible since we dropped TS < 4.1: 
`User.findOne({ where: { '$nested.column$': 'blah2' } })` is not reported as an error by TypeScript anymore.

Closes https://github.com/sequelize/sequelize/issues/13288

### TODO

- [x] Check whether `$name$` is valid
- [x] Check whether `$include.include.name$` is valid
- [x] Add tests for nested nested include

### CODE REVIEW

**Comment by ephys** [comment]
TypeScript < 4.4 does not support using Template Literal Types as keys so I had to ts-ignore both the test & the implementation - but it doesn't cause issues in ts < 4.4 and works in ts >= 4.4

We can enable it again on 2022-05-26 when we drop support for TS < 4.4

**Comment by ephys** [comment]
I think with some more black magic we should be able to typecheck `include.where` too, but I'll keep it for a follow-up PR

```typescript
User.findOne({
  include: [{
    association: User.associations.group,
    where: {
      // these are not currently type-checked
      test: 'test',
    }
  }],
});
```

**Comment by ephys** [comment]
This PR should also include support for the JSON nested key syntax https://sequelize.org/v4/manual/tutorial/querying.html#nested-key

I can't find the documentation on that syntax for v6. Is it still a thing?

**Comment by WikiRik** [comment]
> This PR should also include support for the JSON nested key syntax https://sequelize.org/v4/manual/tutorial/querying.html#nested-key
> 
> I can't find the documentation on that syntax for v6. Is it still a thing?

It seems that it got lost in https://github.com/sequelize/sequelize/pull/11825

**Comment by ephys** [comment]
I'll open a follow-up PR for the `json.syntax` as the typing I have in mind are a bit complicated

File: types/lib/model.d.ts
Comment (AllAwesome497) [inline types/lib/model.d.ts:387]: won't this still be broken in ts <4.4?

It might be better to do something like `& { [attribute: string]: WhereValue<any>` (iirc it'll keep the other keys typed but anything else will be accepted, can't recall for certain though)

Comment (ephys) [reply types/lib/model.d.ts:387]: It will still be broken in TS < 4.4, but not more than before (it's just going to ignore this addition)

Strict typings in `where` is so important IMO that it's not worth it to dumb it down for old TS versions.
I've caught a handful of broken `where` in my own code after properly typing my attributes.

If someone wants to fix this in TS < 4.4, I'd say open a follow-up PR that uses [`typesVersion`](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#version-selection-with-typesversions) to provide TS < 4.4 specific code.

**Comment by ephys** [comment]
As the only issue is that the false-negative is still present in TS < 4.4, I think I'll merge this. Then we can fix https://github.com/sequelize/sequelize/issues/14038, and then open a follow-up PR  that uses `typesVersion` to provide an alternative typing for TS < 4.4

**Comment by WikiRik** [comment]
Sounds good! We should also publish a new alpha release of v7 after we merged all these typing PRs

**Comment by ephys** [comment]
@WikiRik @AllAwesome497 if you're ok with the above plan, can I get one of your review approvals? :)

**Review by WikiRik — APPROVED** [review]

**Pull request merged**
